File: VRTMediationAdapter.m
#import "VRTMediationAdapter.h"
#import <VrtcalSDK/VrtcalSDK.h>

// Initialization configuration keys
static NSString * const keyAppId = @"app_id";

typedef void(^VRTAdapterInitializationCompletionHandler)(MAAdapterInitializationStatus initializationStatus, NSString *_Nullable errorMessage);

@interface VRTMediationAdapter() <MAAdViewAdapter, MAInterstitialAdapter>
@property VRTAdapterInitializationCompletionHandler completionHandler;


@property id <MAAdViewAdapterDelegate> bannerDelegate;

@property id <MAInterstitialAdapterDelegate> interstitialDelegate;

@end

@implementation VRTMediationAdapter

#pragma mark - ALMediationAdapter
- (void)initializeWithParameters:(nonnull id<MAAdapterInitializationParameters>)parameters withCompletionHandler:(nonnull void (^)(void))completionHandler {
    VRTLogWhereAmI();
}


- (void)initializeWithParameters:(id<MAAdapterInitializationParameters>)parameters completionHandler:(void(^)(MAAdapterInitializationStatus initializationStatus, NSString *_Nullable errorMessage))completionHandler {
    
    NSLog(@"initializeWithParameters");
    self.completionHandler = completionHandler;
    
    NSString *strAppId = parameters.serverParameters[keyAppId];
    if (!strAppId || ![strAppId isKindOfClass:[NSString class]]) {
        self.completionHandler(MAAdapterInitializationStatusInitializedFailure, @"Unable to extract appId");
        return;
    }
    
    int appId = [strAppId intValue];
    dispatch_async(dispatch_get_main_queue(), ^{
        [VrtcalSDK initializeSdkWithAppId:appId sdkDelegate:self];
        self.completionHandler(MAAdapterInitializationStatusInitializing, nil);
    });
}

- (void)sdkInitializationFailedWithError:(nonnull NSError *)error {
    NSLog(@"initializeWithParameters");
    self.completionHandler(MAAdapterInitializationStatusInitializedFailure, [error description]);
}

- (void)sdkInitialized {
    VRTLogWhereAmI();
    self.completionHandler(MAAdapterInitializationStatusInitializedSuccess, nil);
}

- (void)destroy {
    VRTLogWhereAmI();
}


#pragma mark - MAAdViewAdapter

- (void)loadAdViewAdForParameters:(nonnull id<MAAdapterResponseParameters>)parameters adFormat:(nonnull MAAdFormat *)adFormat andNotify:(nonnull id<MAAdViewAdapterDelegate>)delegate {
    VRTLogWhereAmI();
    self.bannerDelegate = delegate;
    

    int zoneId = [parameters.thirdPartyAdPlacementIdentifier intValue];
    if (zoneId <= 0) {
        NSString *message = [NSString stringWithFormat:@"Could not extract zoneId from parameters: %@", parameters];
        MAAdapterError *error = [MAAdapterError errorWithCode:MAAdapterError.errorCodeInvalidConfiguration errorString:message];
        [self.bannerDelegate didFailToLoadAdViewAdWithError:error];
        return;
    }

    self.vrtBanner = [[VRTBanner alloc] initWithFrame:CGRectMake(0, 0, adFormat.size.width, adFormat.size.height)];
    self.vrtBanner.adDelegate = self;
    [self.vrtBanner loadAd:zoneId];
}







#pragma mark - MAInterstitialAdapter

- (void)loadInterstitialAdForParameters:(nonnull id<MAAdapterResponseParameters>)parameters andNotify:(nonnull id<MAInterstitialAdapterDelegate>)delegate {
    VRTLogWhereAmI();
    
    self.interstitialDelegate = delegate;
    
    int zoneId = [parameters.thirdPartyAdPlacementIdentifier intValue];
    if (zoneId <= 0) {
        NSString *message = [NSString stringWithFormat:@"Could not extract zoneId from parameters: %@", parameters];
        MAAdapterError *error = [MAAdapterError errorWithCode:MAAdapterError.errorCodeInvalidConfiguration errorString:message];
        [self.bannerDelegate didFailToLoadAdViewAdWithError:error];
        return;
    }
    
    self.vrtInterstitial = [[VRTInterstitial alloc] init];
    self.vrtInterstitial.adDelegate = self;
    [self.vrtInterstitial loadAd:zoneId];
}

- (void)showInterstitialAdForParameters:(nonnull id<MAAdapterResponseParameters>)parameters andNotify:(nonnull id<MAInterstitialAdapterDelegate>)delegate {
    VRTLogWhereAmI();
    [self.vrtInterstitial showAd];
}


#pragma mark - VRTInterstitialDelegate

- (void)vrtInterstitialAdClicked:(nonnull VRTInterstitial *)vrtInterstitial {
    VRTLogWhereAmI();
    [self.interstitialDelegate didClickInterstitialAd];
}


- (void)vrtInterstitialAdDidDismiss:(nonnull VRTInterstitial *)vrtInterstitial {
    VRTLogWhereAmI();
    [self.interstitialDelegate didHideInterstitialAd];
}


- (void)vrtInterstitialAdDidShow:(nonnull VRTInterstitial *)vrtInterstitial {
    VRTLogWhereAmI();
    [self.interstitialDelegate didDisplayInterstitialAd];
}


- (void)vrtInterstitialAdFailedToLoad:(nonnull VRTInterstitial *)vrtInterstitial error:(nonnull NSError *)error {
    VRTLogWhereAmI();
    [self.interstitialDelegate didFailToLoadInterstitialAdWithError:[MAAdapterError errorWithNSError:error]];
}


- (void)vrtInterstitialAdFailedToShow:(nonnull VRTInterstitial *)vrtInterstitial error:(nonnull NSError *)error {
    VRTLogWhereAmI();
    [self.interstitialDelegate didFailToDisplayInterstitialAdWithError:[MAAdapterError errorWithNSError:error]];
}


- (void)vrtInterstitialAdLoaded:(nonnull VRTInterstitial *)vrtInterstitial {
    [self.interstitialDelegate didLoadInterstitialAd];
}


- (void)vrtInterstitialAdWillDismiss:(nonnull VRTInterstitial *)vrtInterstitial {
    VRTLogWhereAmI();
    // No Analog
}


- (void)vrtInterstitialAdWillLeaveApplication:(nonnull VRTInterstitial *)vrtInterstitial {
    VRTLogWhereAmI();
    // No Analog
}


- (void)vrtInterstitialAdWillShow:(nonnull VRTInterstitial *)vrtInterstitial {
    VRTLogWhereAmI();
    // No Analog
}


- (void)vrtInterstitialVideoCompleted:(nonnull VRTInterstitial *)vrtInterstitial {
    VRTLogWhereAmI();
    // No Analog
}


- (void)vrtInterstitialVideoStarted:(nonnull VRTInterstitial *)vrtInterstitial {
    VRTLogWhereAmI();
    // No Analog
}



@end

File: VRTMediationAdapter.swift
//  Converted to Swift 5.8.1 by Swiftify v5.8.26605 - https://swiftify.com/
import AppLovinSDK

// Initialization configuration keys
private let keyAppId = "app_id"
typealias VRTAdapterInitializationCompletionHandler = (MAAdapterInitializationStatus, String?) -> Void

class VRTMediationAdapter: ALMediationAdapter, MAAdViewAdapter, VrtcalSdkDelegate, VRTBannerDelegate, MAInterstitialAdapter, VRTInterstitialDelegate {
    private var completionHandler: VRTAdapterInitializationCompletionHandler?
    private var vrtBanner: VRTBanner?
    private var bannerDelegate: MAAdViewAdapterDelegate?
    private var vrtInterstitial: VRTInterstitial?
    private var interstitialDelegate: MAInterstitialAdapterDelegate?

    // MARK: - ALMediationAdapter

    func initialize(with parameters: MAAdapterInitializationParameters, withCompletionHandler completionHandler: @escaping () -> Void) {
        VRTLogWhereAmI()
    }

    func initialize(with parameters: MAAdapterInitializationParameters?, completionHandler: @escaping (_ initializationStatus: MAAdapterInitializationStatus, _ errorMessage: String?) -> Void) {

        VRTLogWhereAmI()
        self.completionHandler = completionHandler

        let strAppId = parameters?.serverParameters[keyAppId] as? String
        if strAppId == nil || !(strAppId is NSString) {
            self.completionHandler(MAAdapterInitializationStatusInitializedFailure, "Unable to extract appId")
            return
        }

        let appId = Int(strAppId ?? "") ?? 0
        DispatchQueue.main.async(execute: { [self] in
            VrtcalSDK.initializeSdk(withAppId: appId, sdkDelegate: self)
            self.completionHandler(MAAdapterInitializationStatusInitializing, nil)
        })
    }

    func sdkInitializationFailedWithError(_ error: Error) {
        VRTLogWhereAmI()
        completionHandler?(MAAdapterInitializationStatusInitializedFailure, error.description())
    }

    func sdkInitialized() {
        VRTLogWhereAmI()
        completionHandler?(MAAdapterInitializationStatusInitializedSuccess, nil)
    }

    func destroy() {
        VRTLogWhereAmI()
    }

    // MARK: - MAAdViewAdapter

    func loadAdViewAd(for parameters: MAAdapterResponseParameters, adFormat: MAAdFormat, andNotify delegate: MAAdViewAdapterDelegate) {
        VRTLogWhereAmI()
        bannerDelegate = delegate


        let zoneId = parameters.thirdPartyAdPlacementIdentifier.intValue
        if zoneId <= 0 {
            let message = "Could not extract zoneId from parameters: \(parameters)"
            let error = MAAdapterError(code: MAAdapterError.errorCodeInvalidConfiguration, errorString: message)
            bannerDelegate?.didFailToLoadAdViewAdWithError(error)
            return
        }

        vrtBanner = VRTBanner(frame: CGRect(x: 0, y: 0, width: adFormat.size.width, height: adFormat.size.height))
        vrtBanner?.adDelegate = self
        vrtBanner?.loadAd(zoneId)
    }

    // MARK: - VRTBannerDelegate

    func vrtBannerAdClicked(_ vrtBanner: VRTBanner) {
        VRTLogWhereAmI()
        bannerDelegate?.didClickAdViewAd()
    }

    func vrtBannerAdFailed(toLoad vrtBanner: VRTBanner, error: Error) {
        VRTLogWhereAmI()
        bannerDelegate?.didFailToLoadAdViewAdWithError(MAAdapterError(nsError: error))
    }

    func vrtBannerAdLoaded(_ vrtBanner: VRTBanner, withAdSize adSize: CGSize) {
        VRTLogWhereAmI()
        bannerDelegate?.didLoadAd(forAdView: vrtBanner)
    }

    func vrtBannerAdWillLeaveApplication(_ vrtBanner: VRTBanner) {
        VRTLogWhereAmI()
        // No Analog
    }

    func vrtBannerDidDismissModal(_ vrtBanner: VRTBanner, of modalType: VRTModalType) {
        VRTLogWhereAmI()
        bannerDelegate?.didCollapseAdViewAd()
    }

    func vrtBannerDidPresentModal(_ vrtBanner: VRTBanner, of modalType: VRTModalType) {
        VRTLogWhereAmI()
        bannerDelegate?.didExpandAdViewAd()
    }

    func vrtBannerVideoCompleted(_ vrtBanner: VRTBanner) {
        VRTLogWhereAmI()
        // No Analog
    }

    func vrtBannerVideoStarted(_ vrtBanner: VRTBanner) {
        VRTLogWhereAmI()
        // No Analog
    }

    func vrtBannerWillDismissModal(_ vrtBanner: VRTBanner, of modalType: VRTModalType) {
        VRTLogWhereAmI()
        // No Analog
    }

    func vrtBannerWillPresentModal(_ vrtBanner: VRTBanner, of modalType: VRTModalType) {
        VRTLogWhereAmI()
        // No Analog
    }

    func vrtViewControllerForModalPresentation() -> UIViewController {
        return ALUtils.topViewControllerFromKeyWindow()
    }

    // MARK: - MAInterstitialAdapter

    func loadInterstitialAd(for parameters: MAAdapterResponseParameters, andNotify delegate: MAInterstitialAdapterDelegate) {
        VRTLogWhereAmI()

        interstitialDelegate = delegate

        let zoneId = parameters.thirdPartyAdPlacementIdentifier.intValue
        if zoneId <= 0 {
            let message = "Could not extract zoneId from parameters: \(parameters)"
            let error = MAAdapterError(code: MAAdapterError.errorCodeInvalidConfiguration, errorString: message)
            bannerDelegate?.didFailToLoadAdViewAdWithError(error)
            return
        }

        vrtInterstitial = VRTInterstitial()
        vrtInterstitial?.adDelegate = self
        vrtInterstitial?.loadAd(zoneId)
    }

    func showInterstitialAd(for parameters: MAAdapterResponseParameters, andNotify delegate: MAInterstitialAdapterDelegate) {
        VRTLogWhereAmI()
        vrtInterstitial?.showAd()
    }

    // MARK: - VRTInterstitialDelegate

    func vrtInterstitialAdClicked(_ vrtInterstitial: VRTInterstitial) {
        VRTLogWhereAmI()
        interstitialDelegate?.didClickInterstitialAd()
    }

    func vrtInterstitialAdDidDismiss(_ vrtInterstitial: VRTInterstitial) {
        VRTLogWhereAmI()
        interstitialDelegate?.didHideInterstitialAd()
    }

    func vrtInterstitialAdDidShow(_ vrtInterstitial: VRTInterstitial) {
        VRTLogWhereAmI()
        interstitialDelegate?.didDisplayInterstitialAd()
    }

    func vrtInterstitialAdFailed(toLoad vrtInterstitial: VRTInterstitial, error: Error) {
        VRTLogWhereAmI()
        interstitialDelegate?.didFailToLoadInterstitialAdWithError(MAAdapterError(nsError: error))
    }

    func vrtInterstitialAdFailed(toShow vrtInterstitial: VRTInterstitial, error: Error) {
        VRTLogWhereAmI()
        interstitialDelegate?.didFailToDisplayInterstitialAdWithError(MAAdapterError(nsError: error))
    }

    func vrtInterstitialAdLoaded(_ vrtInterstitial: VRTInterstitial) {
        interstitialDelegate?.didLoadInterstitialAd()
    }

    func vrtInterstitialAdWillDismiss(_ vrtInterstitial: VRTInterstitial) {
        VRTLogWhereAmI()
        // No Analog
    }

    func vrtInterstitialAdWillLeaveApplication(_ vrtInterstitial: VRTInterstitial) {
        VRTLogWhereAmI()
        // No Analog
    }

    func vrtInterstitialAdWillShow(_ vrtInterstitial: VRTInterstitial) {
        VRTLogWhereAmI()
        // No Analog
    }

    func vrtInterstitialVideoCompleted(_ vrtInterstitial: VRTInterstitial) {
        VRTLogWhereAmI()
        // No Analog
    }

    func vrtInterstitialVideoStarted(_ vrtInterstitial: VRTInterstitial) {
        VRTLogWhereAmI()
        // No Analog
    }
}